src/basic/alloc-util.h

```diff
@@ -184,17 +185,32 @@ void* greedy_realloc0(void **p, size_t need, size_t size);
 #  define msan_unpoison(r, s)
 #endif
 
-/* This returns the number of usable bytes in a malloc()ed region as per malloc_usable_size(), in a way that
- * is compatible with _FORTIFY_SOURCES. If _FORTIFY_SOURCES is used many memory operations will take the
- * object size as returned by __builtin_object_size() into account. Hence, let's return the smaller size of
- * malloc_usable_size() and __builtin_object_size() here, so that we definitely operate in safe territory by
- * both the compiler's and libc's standards. Note that __builtin_object_size() evaluates to SIZE_MAX if the
- * size cannot be determined, hence the MIN() expression should be safe with dynamically sized memory,
- * too. Moreover, when NULL is passed malloc_usable_size() is documented to return zero, and
- * __builtin_object_size() returns SIZE_MAX too, hence we also return a sensible value of 0 in this corner
- * case. */
-#define MALLOC_SIZEOF_SAFE(x) \
-        MIN(malloc_usable_size(x), __builtin_object_size(x, 0))
+void *expand_to_usable(void *p, size_t newsize) _malloc_ _alloc_(2) _returns_nonnull_;
+
+static inline size_t malloc_sizeof_safe(void **xp) {
+        if (__builtin_expect(!xp || !*xp, 0))
```

Comment:
The `__builtin_expect` would give a hint to the compiler that the `true` branch is unexpected so that it can arrange the generated assembly accordingly.  That is, we'd like to optimize for cases when the passed pointer is a valid live one, not when it is NULL.

`assert` on the other hand will abort if the input condition is false.

I can remove the `__builtin_expect` if such hints are seen as unnecessary in this context.

Reply:
I see. But `xp` cannot and should not be `NULL` here, right? Then, at least I prefer `assert()` for `xp` here. Note, in `assert()` we use `_unlikely_`.

For `*xp`, using `__builtin_expect()` sounds OK for me.

But these are not strong opinion.

Reply:
So it could be in case of, e.g. this test in `src/test/test-alloc-util.c`:

```
assert_se(MALLOC_ELEMENTSOF((float*) NULL) == 0); 
assert_se(MALLOC_SIZEOF_SAFE((float*) NULL) == 0);
```

In this case (or for that matter in any case where the input to the macro is a constant), the `__builtin_choose_expr` will choose `NULL` for `xp`. In fact, `*xp` will be `NULL` (and `xp` non-NULL) only in cases that the compiler is unable to deduce that the input pointer `x` is `NULL`.